contributors/design-proposals/thirdpartyresources.md

```diff
@@ -0,0 +1,224 @@
+proposal covers the steps we believe are necessary to move TPR to beta and to
+prevent future challenges in upgrading.
+
+
+## Goals
+1. Ensure ThirdPartyResource authors can version their APIs and offer their
+users a safe migration path between versions (including alpha->beta for the
+TPR, but also including a discussion of how a TPR author could perform a
+migration for their end users)
+2. Enable ThirdPartyResources to specify how they will appear in API
+discovery to be consistent with other resources and avoid naming confilcts
+3. Move TPR into their own API group to allow the extensions group to be
+[removed](https://github.com/kubernetes/kubernetes/issues/43214)
+4. Support cluster scoped TPR resources
+5. Identify other features required for TPR to become beta
+
+Non-goals
+1. Solve automatic conversion of TPR between versions or automatic migration of
+existing TPR
+
+### Desired API Semantics
+TPRs are intended to look like normal kube-like resources to external clients.
+In order to do that effectively, they should respect the normal get, list,
+watch, create, patch, update, and delete semantics.
+
+In "normal" Kubernetes APIs, if I have a persisted resource in the same group
+with the same name in v1 and v2, they are backed by the same underlying object.
+A change made to one is reflected in the other. API clients, garbage collection,
+namespace cleanup, version negotiation, and controller all build on this.
+
+The convertibility of Kubernetes APIs provides a seamless interaction between
+versions.  A TPR does not have the ability to convert between versions, so this
+experience is broken.
+
+Allowing a single, user specified version for a given TPR will provide this
+semantic by preventing server-side versioning altogether.
+
+
+### Avoiding Naming Problems
+There are several identifiers that a Kubernetes API resource has which share
+value-spaces within an API group and must not conflict.  They are:
+1. Resource-type value space
+  1. plural resource-type name - like "configmaps"
+  2. singular resource-type name - like "configmap"
+  3. short names - like "cm"
+2. Kind-type value space
+  1. Kind name - like "ConfigMap"
+  2. ListKind name - like "ConfigMapList"
+If these values conflict within their value-spaces then no client will be able
+to properly distinguish intent.
+
+The actual name of of the TPR-registration (resource that describes the TPR to
+create) resource can only protect one of these values from conflict.  Since
+Kubernetes API types are accessed via a URL that looks like `/apis/<group>/<version>/namespaces/<namespace-name>/<plural-resource-type>`,
+the name of the TPR-registration object will be `<plural-resource-type>.<group>`.
+
+Conflicts with other parts of the value-space can not be detected with static
+validation, so there will be a spec/status split with `status.conditions` that
+reflect the acceptance status of a TPR-registration.
```

Comment:
"static validation" here means that you want to atomically derive seriver-side whether a new TPR is conflict free and can be created. But this only works via encoding everything uniquely into a TPR name, leading to etcd conflicts. Do I get this correctly?

So to avoid this problem we turn the whole validation into a dynamic status-condition.

In the world of one apiserver this feels overcomplicated, i.e. with some locking magic in etcd one could enforce concistency of all resources on TPR creation (which happens not very often anyway). Would that solve the issue?

This breaks down of course if there are multiple aggregated apiservers. Even in that world you status.condition would only consider a local view, would it?